### DISCUSSION TITLE
V2 equivalent to `session.query(User, UserImage).join(UserImage).all()`?

### DISCUSSION THREAD

**Discussion by solzard** [sqlalchemy/sqlalchemy#9529] [Usage Questions] [answered]
TLDR: What's equivalent to `Session.query().join()` in the V2 (`Session.execute(select()...)`) style?

I'm trying to rewrite V1-styled queries to V2 ones on a `flask-sqlalchemy` system.

The following V1 query 
`db.session.query(User, UserImage).join(UserImage).all()`
returns 
`[(<User 1>, <UserImage 1>), (<User 1>, <UserImage 2>), (<User 2>, <UserImage 3>)]`

I thought 
`db.session.execute(select(User, UserImage).join(UserImage)).scalars().all()`
is an equivalent query in V2; however, got a different result
`[<User 1>, <User 1>, <User 2>]`

Where should I fix it to get the same result?

```python
from flask_login import UserMixin
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()


class User(db.Model, UserMixin):  # type: ignore
    __tablename__ = "users"
    id = db.Column(db.Integer, primary_key=True)  # noqa: A003
    user_images = db.relationship(
        "UserImage", backref="user", order_by="desc(UserImage.id)"
    )


class UserImage(db.Model):  # type: ignore
    __tablename__ = "user_images"
    id = db.Column(db.Integer, primary_key=True)  # noqa: A003
    user_id = db.Column(db.Integer, db.ForeignKey("users.id"))
```

**Comment by zzzeek** [comment, marked as answer]
> I thought
db.session.execute(select(User, UserImage).join(UserImage)).scalars().all()
is an equivalent query in V2; however, got a different result
[<User 1>, <User 1>, <User 2>]

`select(User, UserImage)` returns rows with two columns each, but you are calling [result.scalars()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.scalars) which will return the first column of each row alone.    if you want rows back, call `execute(query).all()` without calling `scalars()`.

**Reply by solzard** [reply]
Thank you for your quick help and great effort in maintaining SQLAlchemy, @zzzeek!
